mantidproject/mantid
# Add error calculation to interpolated points in absorption corrections

**Description of work.**

A previous change (#28979) added an error calculation into the MonteCarloAbsorption algorithm. This PR extends that work to also generate an error for any interpolated points. The Monte Carlo absorption calculation is able to perform the following types of interpolation:

1) wavelength interpolation within a single spectrum using either linear or cubic spline interpolation. This is required in cases where the simulation only generates absorption correction factors at a subset of the wavelength points in the spectrum (controlled via the NumberOfWavelengthPoints parameter in the MonteCarloAbsorption algorithm)

2) spatial interpolation of entire histograms between spectra. It previously used an inverse weighted distance method but it has been changed to use bilinear interpolation as part of this PR because the interpolation error is easier to calculate. This feature is referred to as "Sparse workspace" in the MonteCarloAbsorption algorithm and it is used on instruments with a v large number of detectors where a subset are simulated and the results for the rest of the detectors are obtained by interpolation. The spatial interpolation often also requires some interpolation of type 1) to be done as a final step because the range\bins of the histograms being interpolated may not line up.

For each type of interpolation the error calculation combines two source of error which are treated as independent (added in quadrature): 
- the errors in the simulated points are propagated through the relevant interpolation formula 
- an estimate of the interpolation error for linear and for spatial interpolation is calculated ie how well does the interpolation match the actual function. An interpolation error hasn't been calculated for the spline interpolation because it is complex to calculate (requires a 4th derivative) and it is expected to be small anyway.

The linear and spline interpolation was previously performed using a mix of native and GSL library functions. This has been changed to use Mantid implementations only because the GSL library didn't do the required error calculations.

*Bilinear interpolation*

The change to bilinear interpolation generates slightly different results to the inverse distance weighted approach (ie "y" changes as well as making it easier to calculate "e"). Both approaches use the 4 detectors in the sparse workspace that form a rectangle around the point where interpolation is requested. The biggest differences occur for points that lie halfway between between two detectors on the edge of the "rectangle". For bilinear interpolation the interpolation only includes contributions from the two detectors lying on that edge whereas the inverse distance weighted approach includes contributions from all 4 detectors still. In the ILLDirectGeometryReduction system test, most of the differences are 1-2% with a small number as large as ~10%. The approach of only using 2 of the 4 points seems more sensible so I have proceeded with this change.

*Future work*

There is an algorithm called SplineInterpolation that also does spline interpolation using the CubicSpline function that is available in the Mantid fitting functionality. This algorithm hasn't been updated as part of this PR because it's not used in the absorption correction functionality but it could be done in future. For some reason the Interpolate.cpp and CubicSpline.cpp units had separate\similar code that called GSL interpolation. This could be centralised to use the new native spline interpolation code written as part of this PR.

**To test:**

Run the following MonteCarloAbsorption calculation. It uses an instrument that has 25 detectors arranged in a square grid and it runs the Monte Carlo simulation on a sparse set of 3 x 3 detectors with data for the remaining 16 detectors being interpolated eg spectra numbers 1, 11, 3, 13 form a square of detectors where the simulation did a full calculation and data for spectrum 7 in the middle will be interpolated. Also within each spectrum only a subset of the wavelength points are simulated with the rest being interpolated.
Check the y values on the spatially interpolated spectra (eg workspace index 6, spectrum 7) look similar to the nearby spectra.
Observe the errors are now calculated for all the bins in all the spectra. If you plot a spectrum with error bars it should be possible to see the errors are slightly larger for the interpolated points.
Change the Interpolation parameter to 'Linear' and check the script works and that errors are still calculated for all bins on all spectra.

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

data = CreateSampleWorkspace(WorkspaceType='Histogram', NumBanks=1, BankPixelWidth=5, XMin=1000)
data = ConvertUnits(data, Target='Wavelength')
SetSample(data, Geometry={'Shape': 'Cylinder', 'Height': 5.0, 'Radius': 1.0,
                  'Center': [0.0,0.0,0.0]},
                Material={'ChemicalFormula': 'V'})

abscor = MonteCarloAbsorption(data, SparseInstrument=True,
                              NumberOfDetectorRows=3, NumberOfDetectorColumns=3,
                              Interpolation='CSpline', # test also with this set to Linear
                              ResimulateTracksForDifferentWavelengths=True,
                              NumberOfWavelengthPoints=5)
corrected = data/abscor
```

Fixes #29058 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
